python/modules/IcePy/Operation.cpp

```diff
@@ -370,12 +358,10 @@ extern "C"
     PyObject* exceptions;
     if (!PyArg_ParseTuple(
             args,
-            STRCAST("sO!O!iOO!O!O!OO!"),
```

Comment:
Are you sure you removed the right on here? You removed two parameters and only one character from the string. I also think possibly from the wrong position.

Reply:
The pattern I picked up on, is that I think each `O!` parameter takes a pair of arguments like (type, value).
So, I broke it down like:
`s`: `&name`
`O!`: `(modeType, &mode)`
`O!`: `(modeType, &sendMode)`
`i`: `&amd`
`O`: `&format`
`O!`: `(&PyTuple_Type, &metaData)`
`O!`: `(&PyTuple_Type, &inParams)`
`O!`: `(&PyTuple_Type, &outParams)`
`O`: `&returnType`
`O!`: `(&PyTuple_Type, &exceptions)`

Again, I have no actual basis for doing that. It's just the first pattern I caught onto.
And it seems consistent, and everything adds up, but feel free to tell me otherwise!

Reply:
Looks correct to me. See https://docs.python.org/3/c-api/arg.html

> O! (object) [typeobject, PyObject *]
Store a Python object in a C object pointer. This is similar to O, but takes two C arguments: the first is the address of a Python type object, the second is the address of the C variable (of type [PyObject](https://docs.python.org/3/c-api/structures.html#c.PyObject)*) into which the object pointer is stored. If the Python object does not have the required type, [TypeError](https://docs.python.org/3/library/exceptions.html#TypeError) is raised.